crates/matrix-sdk-crypto/src/store/mod.rs

```diff
@@ -126,6 +127,41 @@ impl<'a> Deref for StoreCacheGuard<'a> {
     }
 }
 
+/// A temporary transaction (that implies a write) to the underlying store.
+#[allow(missing_debug_implementations)]
+pub struct StoreTransaction {
+    store: Store,
+    changes: PendingChanges,
+    // TODO hold onto the cross-process crypto store lock + cache.
```

Comment:
Old habits die hard.

```suggestion
    // TODO: hold onto the cross-process crypto store lock + cache.
```

I'm not sure if I'm completely enthusiastic about this approach compared to having a real transaction exposed by the store itself.

We're implementing a transaction with a `Changes` struct and the cross-process lock. Why not use a real transaction instead?

I'm probably biased because we already have faux transactions with the `Changes` struct and I probably dislike what I implemented quite a fair bit, so happy to be convinced otherwise.

Reply:
So we discussed this, and the reasoning was the following:

- having a transaction associated type in the `CryptoStore` trait would likely duplicate a bit of logic, so having something higher-level made sense.
- a store-level transaction would need to allow read transactions in the middle of write transactions. Unfortunately in the context of indexeddb, any `await` point can cause the browser to flush pending changes to an indexeddb store, invalidating the whole point of the transaction. So, because of indexeddb, we need to queue all the changes, and perform them at a later point in time, in an actual atomic (non-async) indexeddb transaction.
- we would need higher-level wrappers for the cross-process lock to be taken too.
- we would need higher-lever wrappers for interactions with the cache too (?).

In light of those arguments, we decided that the current approach (more deeply commented in the PR description) was fine.